### PR TITLE
ARROW 13473: [C++] Add super-scalar impl for BitUtil::SetBitTo

### PR DESCRIPTION
**Pull request by gcca**
Add the super-scalar variant to set a bit.

### CODE REVIEW

**Comment by cyb70289** [comment]
Do you have benchmark results showing this code is faster than original one?
Compilers are pretty smart in optimizing these common code patterns. Without solid benchmark data, I would prefer simpler code.

Disassembler shows the dumb `if else` actually generates almost same code as this bithack trick. Modern CPUs all have `conditional move` instructions to remove trivial branches.
https://godbolt.org/z/fqzs7rGev

**Comment by nirandaperera** [comment]
@gcca the title should be "ARROW-13473: [C++] Add super-scalar impl for BitUtil::SetBitTo"

**Comment by nirandaperera** [comment]
@cyb70289 this was discussed in this PR #10679. The idea came from @wesm. 
Ref: https://graphics.stanford.edu/~seander/bithacks.html#ConditionalSetOrClearBitsWithoutBranching

The idea is to check if these two methods have any performance implications. 
```c++
void SetBitTo(uint8_t* bits, int64_t i, bool bit_is_set) {
  bits[i / 8] ^= static_cast<uint8_t>(-static_cast<uint8_t>(bit_is_set) ^ bits[i / 8]) & kBitmask[i % 8];
}

void SetBitSuperscalarTo(uint8_t* bits, int64_t i, bool bit_is_set) {
  bits[i / 8] = (bits[i / 8] & ~kBitmask[i % 8]) | (-static_cast<uint8_t>(bit_is_set) & kBitmask[i % 8]);
}
```

**Comment by pitrou** [comment]
@github-actions crossbow submit *valgrind*

**Comment by pitrou** [comment]
@ursabot please benchmark

**Comment by pitrou** [comment]
@github-actions crossbow submit conda-cpp-valgrind

File: cpp/src/arrow/util/bit_util.h [hunk 1]
```diff
@@ -310,8 +310,13 @@ static inline void SetBitTo(uint8_t* bits, int64_t i, bool bit_is_set) {
   // "Conditionally set or clear bits without branching"
   // NOTE: this seems to confuse Valgrind as it reads from potentially
   // uninitialized memory
+#ifdef SUPERSCALAR_CPU
```
Comment (pitrou) [inline cpp/src/arrow/util/bit_util.h:313]: Where does this come from?

**Comment by gcca** [comment]
@nirandaperera: these are the results when running the benchmarks

**Current version**
|Benchmark           |                            Time          |   CPU  | Iterations |UserCounters |
|------------------------|---------------------------------------|----------|--------------|----------|
|BenchmarkBitmapVisitBitsetAnd/32768/0  |  57776803 ns   |  57772660 ns     |      11 | bytes_per_second=553.895k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/0 | 232098279 ns  |  232094670 ns   |         3 | bytes_per_second=551.499k/s|
|BenchmarkBitmapVisitBitsetAnd/32768/0  |  57848824 ns   |  57843817 ns       |    12|  bytes_per_second=553.214k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/0 |  234126020 ns  |  234115416 ns    |        3 |bytes_per_second=546.739k/s|
|BenchmarkBitmapVisitBitsetAnd/32768/1  |  61148854 ns    | 61148534 ns       |    10| bytes_per_second=523.316k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/1 | 244266155 ns  |  244259941 ns    |        3 |bytes_per_second=524.032k/s|
|BenchmarkBitmapVisitBitsetAnd/32768/2  |  61629097 ns   |  61623725 ns       |    11 |bytes_per_second=519.281k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/2 | 247708290 ns |   247696540 ns     |       3| bytes_per_second=516.761k/s|

**Superscalar CPU version**
|Benchmark               |                        Time        |     CPU |  Iterations |UserCounters|
|----------------------------|---------------------------------|------------|--------------|----------|
|BenchmarkBitmapVisitBitsetAnd/32768/0    |58131617 ns    | 58130347 ns   |        12 |bytes_per_second=550.487k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/0  |233650397 ns   | 233636376 ns |           3| bytes_per_second=547.86k/s|
|BenchmarkBitmapVisitBitsetAnd/32768/0    |58134623 ns    | 58133313 ns     |      12 |bytes_per_second=550.459k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/0  |232643208 ns   | 232640618 ns  |          3| bytes_per_second=550.205k/s|
|BenchmarkBitmapVisitBitsetAnd/32768/1   | 61623058 ns    | 61614097 ns    |       11| bytes_per_second=519.362k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/1  |251259627 ns   | 251242745 ns  |          3 |bytes_per_second=509.467k/s|
|BenchmarkBitmapVisitBitsetAnd/32768/2    |62266910 ns    | 62259970 ns      |     12| bytes_per_second=513.974k/s|
|BenchmarkBitmapVisitBitsetAnd/131072/2 | 246426455 ns |   246412586 ns   |         3| bytes_per_second=519.454k/s|

**Comment by nirandaperera** [comment]
@ursabot please benchmark

**Comment by pitrou** [comment]
This doesn't seem very conclusive. Recommend closing.

**Comment by cyb70289** [comment]
Benchmark result and static analysis both show this patch is not necessary for modern compiler and hardware (I do believe bithack tricks are very useful and interesting to inspire practical algorithms).

Can we close this pr and related jira card? @gcca @nirandaperera 
And welcome more contributions.

**Comment by pitrou** [comment]
Agreed with @cyb70289 that this doesn't seem to show promise. Closing.

**Pull request closed without merging**